easybuild/framework/easyblock.py

```diff
@@ -2720,20 +2720,66 @@ def prepare_step(self, start_dir=True, load_tc_deps_modules=True):
         if start_dir:
             self.guess_start_dir()
 
+    def _run_command_stack(self, commands, caller):
+        """Execute stack of commands in a shell"""
+        self.log.debug(f"Stack of commands to be executed: {commands}")
+
+        # make sure we have a list of commands
+        if not isinstance(commands, (list, tuple)):
+            error_msg = f"Invalid value for '{caller}', should be list or tuple of strings: {commands}"
+            raise EasyBuildError(error_msg)
+
+        for cmd in commands:
+            if not isinstance(cmd, str):
+                raise EasyBuildError(f"Invalid element in '{caller}', not a string: {cmd}")
+            run_shell_cmd(cmd)
+
+    def run_step_main_action(self, action, step_name, pre_cmds=None, pre_opts="", post_opts=""):
+        """Construct main command of step and execute it"""
+        if pre_cmds:
+            self._run_command_stack(pre_cmds, f"pre_{step_name}_cmds")
+
+        step_cmd = f"{pre_opts} {action} {post_opts}"
+        return run_shell_cmd(step_cmd)
+
     def configure_step(self):
-        """Configure build  (abstract method)."""
-        raise NotImplementedError
+        """Configure build."""
+        if self.cfg['configure_cmd'] is not None:
+            res = self.run_step_main_action(
+                self.cfg['configure_cmd'],
+                "configure",
+                self.cfg['pre_configure_cmds'],
+                self.cfg['preconfigopts'],
+                self.cfg['configopts'],
+            )
+            return res.output
 
     def build_step(self):
-        """Build software  (abstract method)."""
-        raise NotImplementedError
+        """Build software."""
+        if self.cfg['build_cmd'] is not None:
+            res = self.run_step_main_action(
+                self.cfg['build_cmd'],
+                "build",
+                self.cfg['pre_build_cmds'],
+                self.cfg['prebuildopts'],
+                self.cfg['buildopts'],
+            )
+            return res.output
 
     def test_step(self):
         """Run unit tests provided by software (if any)."""
-        unit_test_cmd = self.cfg['runtest']
-        if unit_test_cmd:
-            self.log.debug(f"Trying to execute {unit_test_cmd} as a command for running unit tests...")
-            res = run_shell_cmd(unit_test_cmd)
+        if self.cfg['runtest'] is not None:
+            self.log.deprecated("Easyconfig parameter 'runtest' is deprecated, use 'test_cmd' instead.", "6.0")
```

Comment:
@lexming This implies updating all easyblocks/easyconfigs to use `test_cmd` instead of `runtest` before we release EasyBuild 5.0, so please open an issue on that in both repos and tag it for EasyBuild 5.0

Reply:
Also, what if both `test_cmd` and `runtest` are set?

Either we error out, or we let one overrule the other (with logging).

Reply:
good point, I suggest to not error out and just ignore `runtest` if both are defined (with a warning ofc). Fixed in https://github.com/easybuilders/easybuild-framework/pull/4531/commits/22d96485b45bcf7abf5e337cfb524cc9a3453d2c